[RFC] Remove all occurrences of the mb_ptr2char macro

ZviRackover opened the pull request.

ZviRackover committed:
Remove all occurences of the mb_ptr2char macro

First step towards implemening issue #7401.

The same can be done for all deprecated mb_ functions in follow-up
patches.

ZviRackover committed:
lint: clean-up after parent commit

justinmk commented:
Thanks for the effort. This patch was already done in https://github.com/neovim/neovim/pull/8513 (`./scripts/vim-patch -l` should not have listed it, assuming you have current Neovim master branch).

But it looks like you found more cases?

ZviRackover commented:
yes, i started using the patch script and after hitting conflicts saw that it was already partially merged. Should i retitle the patch?

justinmk commented:
This actually isn't the "spirit" of vim patch 8.0.0520, but more related to https://github.com/neovim/neovim/issues/7401 .

Because `mb_*` is not a function pointer at all in Nvim. 

I see you marked it WIP, but this change should be done in small steps. So if there's no known technical problem with your PR , I would mark it as RFC.

justinmk commented:
@bfredl @ZyX-I Have you done something similar in existing PRs, which may conflict with this? Should we wait on this to avoid conflicts?

ZviRackover commented:
I agree that #7401 is a better match. Will retitle and update commit message.

ZviRackover commented:
@bfredl ok will try to remove `enc_utf8` and `has_mbyte .

ZviRackover committed:
Remove some occrrences of enc_utf8 and has_mbyte

Removing uses and related dead code in the locallity of changes of
the two parent commits.

ZyX-I commented:
I did not do anything to remove all of some macros usages, but if I edit a code with `mb_…` I usually remove it and use `utf*` variant instead.

ZviRackover commented:
This is my first pr for this project and i'm not sure what to do about the failed checks.

- QuickBuild report FAILED but the report did not mention any failed tesrs
- Some of the travis-ci jobs seemed to have failed due to environment issues

KillTheMule commented:
QB: Click on "Tap report" on the left menu, then cycle through the builds by using the dropdown on the right hand side besides the "export" button. You will find some [test failures](http://neovim-qb.szakmeister.net/build/18464/tap_report/failed_tests?reportset=functionaltest-macosx-64-release). They are known and can be ignored by you.

Travis: Looks unrelated indeed, but I've never seen those before. @jamessan or @justinmk could maybe have a look.

jamessan commented:
Those are strange.  I've restarted the Travis job to see if it was a temporary issue.

jamessan commented on src/nvim/regexp.c:
```diff
@@ -6782,19 +6750,20 @@ static int vim_regsub_both(char_u *source, typval_T *expr, char_u *dest,
-        else     /* just copy */
-          cc = c;
-
+        } else {
```
Why was this changed into a plain `} else {` rather than kept as the `} else if (func_all != NULL) {`?

ZyX-I replied:
@ZviRackover Linter is completely fine with the form @jamessan suggested. You need to fix linter errors, but there is no reason to use `else { if }` form in place of `else if`, you just needed to add braces after the `if (cond)`.

ZviRackover replied:
Done in [4f3ce52](https://github.com/neovim/neovim/pull/8665/commits/4f3ce526e644628b200257b4bcf867eb438f8c12])

jamessan commented on src/nvim/spell.c:
```diff
@@ -4292,23 +4289,21 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
-                                                        + sp->ts_fcharstart))) {
+                                            utf_ptr2char(tword + sp->ts_twordlen
+                                                         - sp->ts_tcharlen),
+                                            utf_ptr2char(fword +
```
The `+` should be on the next line if this has to be split.

ZviRackover replied:
Done in latest commit. Note that the ) { were move to a new line to stay within line length limits. Hope they were aligned correctly.

jamessan commented on src/nvim/spell.c:
```diff
@@ -4507,40 +4497,32 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
         sp->ts_state = STATE_REP_INI;
         break;
-      }
+          }
```
Indentation is wrong from here to the `#ifdef DEBUG_TRIEWALK`.

ZviRackover replied:
Done.

oni-link replied:
The new change shits the code to the right, the old indentation was okay.

ZviRackover replied:
Done

jamessan reviewed (commented):
Nice work.  Just a few small things.

janlazo commented:
Will break https://github.com/neovim/neovim/pull/8635 because it uses `mb_ptr2char`.

jamessan commented:
> Will break #8635 because it uses mb_ptr2char.

So why don't use use `utf_ptr2char`?  `mb_ptr2char` is just a `#define`.

ZviRackover committed:
style: fixing minor issues noted in code review.

ZviRackover commented:
The latest push should address all comments from @jamessan

ZyX-I commented on src/nvim/spell.c:
```diff
@@ -4292,23 +4290,22 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
+                                                         - sp->ts_tcharlen),
+                                            utf_ptr2char(fword
+                                                         + sp->ts_fcharstart))
+                           ) {
```
You should not move `)` on its own line. As far as I understand style guide you need to resolve the situation by either moving `!soundfold` (the very *first* part of condition) on its own line and indent it with +4 spaces or move `slang` (the very *first* argument) on its own line with +4 spaces. Following parts of condition or arguments should be aligned with the first one (like now).

ZviRackover replied:
Ok, i'll go for moving `!soundfold` to a new line. If all fails, there is a possibility of refactoring out the expression `tword + sp->ts_twordlen - sp->ts_tcharlen` as it repeats in neighboring conditions

ZviRackover replied:
Here's what can be done following what you siggest @ZyX-I. Only problem is that the rightmost brace reaches column 81 and the linter complains only about that.
Is there another possiblity for fitting this code?
```
                } else if (!soundfold
                           && slang->sl_has_map
                           && similar_chars(slang,
                                            utf_ptr2char(tword
                                                         + sp->ts_twordlen
                                                         - sp->ts_tcharlen),
                                            utf_ptr2char(fword
                                                         + sp->ts_fcharstart))) {
                  // For a similar character adjust score from
                  // SCORE_SUBST to SCORE_SIMILAR.
                  sp->ts_score -= SCORE_SUBST - SCORE_SIMILAR;
                }
```

ZyX-I replied:
@ZviRackover It is not what I suggested and I do not see `!soundfold` on a new line there. What I suggested is

```C
                } else if (
                    !soundfold
                    && slang->sl_has_map
                    && similar_chars(slang,
                                     utf_ptr2char(tword
                                                  + sp->ts_twordlen
                                                  - sp->ts_tcharlen),
                                     utf_ptr2char(fword
                                                  + sp->ts_fcharstart))) {
                  // For a similar character adjust score from
                  // SCORE_SUBST to SCORE_SIMILAR.
                  sp->ts_score -= SCORE_SUBST - SCORE_SIMILAR;
                }
```

Same thing may be done with `slang` as well, I think doing both would be cleaner:

```C
                } else if (
                    !soundfold
                    && slang->sl_has_map
                    && similar_chars(
                        slang,
                        utf_ptr2char(tword + sp->ts_twordlen - sp->ts_tcharlen),
                        utf_ptr2char(fword + sp->ts_fcharstart))) {
                  // For a similar character adjust score from
                  // SCORE_SUBST to SCORE_SIMILAR.
                  sp->ts_score -= SCORE_SUBST - SCORE_SIMILAR;
                }
```

ZviRackover replied:
Thanks. Now i see what you mean.

ZyX-I commented on src/nvim/spell.c:
```diff
@@ -4292,23 +4290,22 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
                   sp->ts_score -= SCORE_SUBST - SCORE_SUBCOMP;
-                } else if (!soundfold
-                           && slang->sl_has_map
+                } else if (!soundfold && slang->sl_has_map
```
Do not join lines like this, lengthy multipart condition is more readable when each part starts on its own line.

ZviRackover commented:
Latest push should address all comments by @ZyX-I

ZviRackover commented:
If there are no additional comments, would it be possible to merge this pr? My concern is that the changes to 25 files in hundreds of lines  are in risk of conflict with other patches. Thanks

oni-link commented on src/nvim/spell.c:
```diff
@@ -3648,7 +3645,8 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
 #ifdef DEBUG_TRIEWALK
   // Stores the name of the change made at each level.
-  char_u changename[MAXWLEN][80];
+  static const size_t changenameElemSize = 80;
```
What is the advantage of the additional `static` here?

ZviRackover replied:
It narrowa the scope of the variable to current file.

oni-link replied:
It narrows it down to function scope and keeps the value between function calls. But why do we want this instead of just a constant value? Is it more efficient?

ZyX-I replied:
@ZviRackover It does not, not in this context. Neither it does “narrow down to function scope”, all variables declared inside function already have only this scope, no matter `static` or not. Just `static` variables declared in a function are essentially *global* variables; you would not get a symbol for them so you can’t just access that outside of function, but you *do* get a constant address which may be returned from the function, one which will continue to be valid after function was finished. This is not needed.

ZyX-I commented on src/nvim/spell.c:
```diff
@@ -3648,7 +3645,8 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
   // Stores the name of the change made at each level.
-  char_u changename[MAXWLEN][80];
+  static const size_t changenameElemSize = 80;
+  char_u changename[MAXWLEN][changenameElemSize];
```
The result looks much like VLA. I am not sure what exactly compiler would do, but you must not define dimensions using anything of which compiler not thinks like it is “an integer constant expression”. According to the standard this construct defenitely *is* a VLA:

> An integer constant expression99) shall have integer type and shall only have operands that are integer constants, enumeration constants, character constants, sizeof expressions whose results are integer constants, and floating constants that are the immediate operands of casts. Cast operators in an integer constant expression shall only convert arithmetic types to integer types, except as part of an operand to the sizeof operator.

though compiler may think otherwise.

Also note that `changenameElemSize` is *not* a valid name for variable according to the style guide. For constants (though I never saw it used for constants scoped to function) it is `kChangenameElemSize`, for variables it is `changename_elem_size`.

And you ultimately not need any variable here at all. I have checked how `void main(){char a[10][20]; printf("%zu", sizeof(a[0]));}` program works and it returns 20 on gcc, clang and tcc, so you should use `sizeof` in place of variable. Also note that comment that is now before `changenameElemSize` definition actually applies to `changename` and *not* `changenameElemSize` which is incorrect.

ZviRackover replied:
I totally missed that this is in a function, so of course the `static` is not needed. Good catch.

ZviRackover commented:
Latest revision removes the redundant `static` discussed above.

ZviRackover commented:
The latest revision removes `changenameElemSize`

ZyX-I commented on src/nvim/spell.c:
```diff
@@ -4542,9 +4525,10 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
-        sprintf(changename[depth], "%.*s-%s: swap %c and %c",
-            sp->ts_twordlen, tword, fword + sp->ts_fidx,
-            c, c2);
+        snprintf(changename[depth], sizeof changename[0],
```
Missing parenthesis: while `sizeof` being operator can be used without them it is always written like if it was a function.

ZviRackover replied:
Done. 
Didn't find this rule in the relevant style guide section, https://neovim.io/develop/style-guide.xml#sizeof. If it is indeed not listed, should it be added?

bfredl commented on src/nvim/regexp.c:
```diff
@@ -1107,13 +1103,9 @@ static int get_coll_element(char_u **pp)
   if (p[0] != NUL && p[1] == '.') {
-    if (has_mbyte)
-      l = (*mb_ptr2len)(p + 2);
+    l = (*mb_ptr2len)(p + 2);
```
`utfc_ptr2len`

ZviRackover replied:
Done

ZviRackover commented:
@justinmk @ZyX-I do you have any additional comments? I see the milestones for this pr are bumped without it being pushed and would like to understand if i am expected to do anything to make progress.

janlazo commented on src/nvim/charset.c:
```diff
@@ -1375,7 +1375,7 @@ void getvvcol(win_T *wp, pos_T *pos, colnr_T *start, colnr_T *cursor,
 
     if (pos->col < (colnr_T)STRLEN(ptr)) {
-      int c = (*mb_ptr2char)(ptr + pos->col);
+      int c = utf_ptr2char(ptr + pos->col);
```
Add `const` unless this changes in a future vim-patch.

ZviRackover replied:
I didn't see any coding style guideline in the wiki requiring `const` on variables which are not pointers.
I don't have a strong objection to adding `const` in cases such as this, but i don't see much benefit in it.

bfredl replied:
Especially with a small local block there is really not much benefit to add const to a local variable binding (it just trivially repeats info already visible in the block), we should focus the effort to pointer targets (where const makes it a distinct type), and possibly bindings in really large scopes.

janlazo commented on src/nvim/edit.c:
```diff
@@ -2311,24 +2311,14 @@ static void ins_compl_longest_match(compl_T *match)
-      if (match->cp_icase ? (mb_tolower(c1) != mb_tolower(c2))
-          : (c1 != c2)) {
+      c1 = utf_ptr2char(p);
+      c2 = utf_ptr2char(s);
```
`c1`, and `c2` are `const` now?

janlazo commented on src/nvim/edit.c:
```diff
@@ -8494,7 +8482,7 @@ int ins_copychar(linenr_T lnum)
     ptr = prev_ptr;
 
-  c = (*mb_ptr2char)(ptr);
+  c = utf_ptr2char(ptr);
```
Is this `const int` now?

janlazo commented on src/nvim/ex_docmd.c:
```diff
@@ -2900,7 +2900,7 @@ const char * set_one_cmd_context(
     while (*p != NUL) {
       if (has_mbyte) {
-        c = mb_ptr2char((const char_u *)p);
+        c = utf_ptr2char((const char_u *)p);
```
If you are keeping only the branch where `has_mbyte` is true, can you be consistent with the rest?

ZviRackover replied:
Good catch, thanks. Will fix all these cases you pointed out.

janlazo commented on src/nvim/mbyte.c:
```diff
@@ -674,7 +674,7 @@ int mb_ptr2char_adv(const char_u **const pp)
   int c;
 
-  c = (*mb_ptr2char)(*pp);
+  c = utf_ptr2char(*pp);
```
`const int c`

janlazo commented on src/nvim/mbyte.c:
```diff
@@ -687,7 +687,7 @@ int mb_cptr2char_adv(const char_u **pp)
   int c;
 
-  c = (*mb_ptr2char)(*pp);
+  c = utf_ptr2char(*pp);
```
`const int c`

janlazo commented on src/nvim/ex_docmd.c:
```diff
@@ -2922,7 +2922,7 @@ const char * set_one_cmd_context(
         while (*p != NUL) {
           if (has_mbyte) {
-            c = mb_ptr2char((const char_u *)p);
+            c = utf_ptr2char((const char_u *)p);
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -3329,8 +3329,9 @@ static bool cmdline_paste(int regname, bool literally, bool remcr)
         if (has_mbyte) {
           len = (*mb_head_off)(ccline.cmdbuff, w - 1) + 1;
-          if (!vim_iswordc(mb_ptr2char(w - len)))
+          if (!vim_iswordc(utf_ptr2char(w - len))) {
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -3841,13 +3842,13 @@ ExpandOne (
       if (has_mbyte) {
         mb_len = (* mb_ptr2len)(&xp->xp_files[0][len]);
-        c0 = (* mb_ptr2char)(&xp->xp_files[0][len]);
+        c0 = utf_ptr2char(&xp->xp_files[0][len]);
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -3841,13 +3842,13 @@ ExpandOne (
       for (i = 1; i < xp->xp_numfiles; ++i) {
         if (has_mbyte) {
-          ci =(* mb_ptr2char)(&xp->xp_files[i][len]);
+          ci = utf_ptr2char(&xp->xp_files[i][len]);
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/fold.c:
```diff
@@ -1787,8 +1787,9 @@ char_u *get_foldtext(win_T *wp, linenr_T lnum, linenr_T lnume,
 
         if (has_mbyte && (len = (*mb_ptr2len)(p)) > 1) {
-          if (!vim_isprintc((*mb_ptr2char)(p)))
+          if (!vim_isprintc(utf_ptr2char(p))) {
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/message.c:
```diff
@@ -2870,8 +2870,9 @@ do_dialog (
       for (i = 0; hotkeys[i]; ++i) {
         if (has_mbyte) {
-          if ((*mb_ptr2char)(hotkeys + i) == c)
+          if (utf_ptr2char(hotkeys + i) == c) {
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/message.c:
```diff
@@ -2912,7 +2913,7 @@ copy_char (
   if (has_mbyte) {
     if (lowercase) {
-      c = mb_tolower((*mb_ptr2char)(from));
+      c = mb_tolower(utf_ptr2char(from));
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/ops.c:
```diff
@@ -3634,10 +3634,10 @@ int do_join(size_t count,
         cend = curr + currsize;
         MB_PTR_BACK(curr, cend);
-        endcurr1 = (*mb_ptr2char)(cend);
+        endcurr1 = utf_ptr2char(cend);
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/option.c:
```diff
@@ -6853,27 +6856,27 @@ void find_mps_values(int *initc, int *findc, int *backwards, int switchit)
       char_u *prev;
 
-      if (mb_ptr2char(ptr) == *initc) {
+      if (utf_ptr2char(ptr) == *initc) {
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/regexp.c:
```diff
@@ -5474,7 +5442,7 @@ regrepeat (
         scan++;
       } else if (has_mbyte && (len = (*mb_ptr2len)(scan)) > 1) {
-        if ((cstrchr(opnd, (*mb_ptr2char)(scan)) == NULL) == testval) {
+        if ((cstrchr(opnd, utf_ptr2char(scan)) == NULL) == testval) {
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/spell.c:
```diff
@@ -4684,7 +4648,7 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
         if (has_mbyte) {
           n = MB_CPTR2LEN(p);
-          c = mb_ptr2char(p);
+          c = utf_ptr2char(p);
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/regexp.c:
```diff
@@ -6878,10 +6846,7 @@ static int vim_regsub_both(char_u *source, typval_T *expr, char_u *dest,
-                  c = mb_ptr2char(s);
-                else
-                  c = *s;
+                c = utf_ptr2char(s);
```
See https://github.com/neovim/neovim/pull/8665/files#r207964834

janlazo commented on src/nvim/spell.c:
```diff
@@ -2609,16 +2611,11 @@ static bool spell_iswordp(char_u *p, win_T *wp)
-      return spell_mb_isword_class(mb_get_class(p), wp);
-    }
-    return spelltab.st_isw[c];
+  int c = utf_ptr2char(p);
```
`const int c`

janlazo commented on src/nvim/regexp_nfa.c:
```diff
@@ -5033,13 +5033,8 @@ static int nfa_regmatch(nfa_regprog_T *prog, nfa_state_T *start,
-      clen = 1;
-    }
+    curc = utf_ptr2char(reginput);
+    clen = utfc_ptr2len(reginput);
```
You can declare and initialize `curc` on same line. Same applies for `clen`.

janlazo commented on src/nvim/regexp.c:
```diff
@@ -3468,10 +3446,7 @@ static long bt_regexec_both(char_u *line,
-      c = (*mb_ptr2char)(prog->regmust);
-    else
-      c = *prog->regmust;
+    c = utf_ptr2char(prog->regmust);
```
`int c = ...`

justinmk commented on src/nvim/misc1.c:
```diff
@@ -1749,11 +1748,7 @@ del_lines (
-  if (has_mbyte)
-    return (*mb_ptr2char)(ptr);
-  return (int)*ptr;
+  return utf_ptr2char(ml_get_pos(pos));
```
Note for next time: it's often good to have a local (`ptr`) anyway, for debugging.
